src/Analyzer/Resolve/IdentifierResolver.cpp

```diff
@@ -407,20 +407,39 @@ QueryTreeNodePtr IdentifierResolver::tryResolveTableIdentifierFromDatabaseCatalo
     }
 
     StorageID storage_id(database_name, table_name);
-    storage_id = context->resolveStorageID(storage_id);
+    storage_id = context->resolveStorageID(storage_id, Context::ResolveAll, /*get_uuid*/ false);
     bool is_temporary_table = storage_id.getDatabaseName() == DatabaseCatalog::TEMPORARY_DATABASE;
 
     StoragePtr storage;
+    TableLockHolder storage_lock;
```

Comment:
Why do we need to lock it? Atomic and Replicated databases never take exclusive locks anyway

Reply:
There was already a `storage->lockForShare` call here. I assumed its purpose is to allow the SELECT to complete if the table was dropped in the middle of it.

> Atomic and Replicated databases never take exclusive locks anyway

Do they have some other mechanism to prevent SELECT from failing if the table was dropped in the middle? Or do SELECTs fail in this case? I tried it and it didn't fail, but maybe it's not intended to be reliable?

Reply:
> There was already a storage->lockForShare call here. I assumed its purpose is to allow the SELECT to complete if the table was dropped in the middle of it.

Yes, it's the case for Ordinary databases, but not for Atomic/Replicated. 

> Do they have some other mechanism to prevent SELECT from failing if the table was dropped in the middle?

Yes, we use refcount in StoragePtr to check if the table is still being used by some queries, we don't really need these locks. However, there's one place where we still rely on the locks: https://github.com/ClickHouse/ClickHouse/blob/a7d0a5991e5e0e0be9bce7bc7ac9e1de5f376903/src/Interpreters/InterpreterSystemQuery.cpp#L873

Maybe we could change that in `tryRestartReplica` and get rid of the locks completely